internal/instrument/instrument.go

```diff
@@ -571,24 +572,85 @@ func funcIdent(e dst.Expr) (*dst.Ident, bool) {
 	return f, true
 }
 
-func useMiddleware(pkg, middleware string) *dst.ExprStmt {
-	stmt := &dst.ExprStmt{
-		X: &dst.CallExpr{
-			Fun: &dst.SelectorExpr{
-				X:   &dst.Ident{Name: pkg},
-				Sel: &dst.Ident{Name: "Use"},
+// useMiddleware returns a statement that uses the given middleware.
+func useMiddleware(expr dst.Expr, middleware string) (*dst.ExprStmt, error) {
+	/*
+		//dd:instrumented
+		r := echo.New()
+		//dd:startinstrument
+		r.Use(instrument.EchoV4Middleware())
+		//dd:endinstrument
+
+		//dd:instrumented
+		api.server = echo.New()
+		//dd:startinstrument
+		api.serverr.Use(instrument.EchoV4Middleware())
+		//dd:endinstrument
+	*/
+	var stmt *dst.ExprStmt
+	switch ex := expr.(type) {
+	case *dst.Ident:
+		stmt = &dst.ExprStmt{
+			X: &dst.CallExpr{
+				Fun: &dst.SelectorExpr{
+					X:   &dst.Ident{Name: ex.Name},
+					Sel: &dst.Ident{Name: "Use"},
+				},
+				Args: []dst.Expr{
+					&dst.CallExpr{
+						Fun: &dst.Ident{
+							Name: middleware,
+							Path: "github.com/datadog/orchestrion/instrument",
+						},
+					},
+				},
 			},
-			Args: []dst.Expr{
-				&dst.CallExpr{
-					Fun: &dst.Ident{
-						Name: middleware,
-						Path: "github.com/datadog/orchestrion/instrument",
+		}
+	case *dst.SelectorExpr:
+		x, ok := ex.X.(*dst.Ident)
+		if !ok {
+			return nil, fmt.Errorf("unexpected type %v", reflect.TypeOf(ex.X))
+		}
+		stmt = &dst.ExprStmt{
+			X: &dst.CallExpr{
+				Fun: &dst.SelectorExpr{
+					X: &dst.SelectorExpr{
+						X:   &dst.Ident{Name: x.Name},
+						Sel: &dst.Ident{Name: ex.Sel.Name},
+					},
+					Sel: &dst.Ident{Name: "Use"},
+				},
+				Args: []dst.Expr{
+					&dst.CallExpr{
+						Fun: &dst.Ident{
+							Name: middleware,
+							Path: "github.com/datadog/orchestrion/instrument",
+						},
```

Comment:
I think we can avoid the duplication by checking first what type of expression we have. Then, in case of a `SelectorExpr`, assign the underlying `Ident`. If not, assign the received expression, always if it's a Ident. From the found expression we can deduce if we need to a `X.Fun.X` `SelectorExpr` equal to `dst.Ident{Name: ex.Name}` or `dst.SelectorExpr{X: dst.Ident{Name: x.Name}, Sel: dst.Ident{Name: ex.Sel.Name}}`.

WDYT, @ahmed-mez?

Reply:
I considered that, but kept the two cases seperated tin the end to ease readability. I agree that we could avoid some duplication but the code would be much harder to read. At least right now the separation between the 2 cases is obvious and the reader would just need to focus on one branch.

It's not a strong opinion though, happy to reconsider. Mind sharing a code suggestion for what you described?

Reply:
@ahmed-mez My suggestion would look like this:

```go
// useMiddleware returns a statement that uses the given middleware.
func useMiddleware(expr dst.Expr, middleware string) (*dst.ExprStmt, error) {
	/*
		//dd:instrumented
		r := echo.New()
		//dd:startinstrument
		r.Use(instrument.EchoV4Middleware())
		//dd:endinstrument

		//dd:instrumented
		api.server = echo.New()
		//dd:startinstrument
		api.server.Use(instrument.EchoV4Middleware())
		//dd:endinstrument
	*/
	var rExpr dst.Expr
	switch ex := expr.(type) {
	case *dst.Ident:
		rExpr = &dst.Ident{Name: ex.Name}
	case *dst.SelectorExpr:
		x, ok := ex.X.(*dst.Ident)
		if !ok {
			return nil, fmt.Errorf("unexpected type %v", reflect.TypeOf(ex.X))
		}
		rExpr = &dst.SelectorExpr{
			X:   &dst.Ident{Name: x.Name},
			Sel: &dst.Ident{Name: ex.Sel.Name},
		}
	default:
		return nil, fmt.Errorf("unexpected type %v", reflect.TypeOf(expr))
	}
	stmt := &dst.ExprStmt{
		X: &dst.CallExpr{
			Fun: &dst.SelectorExpr{
				X:   rExpr,
				Sel: &dst.Ident{Name: "Use"},
			},
			Args: []dst.Expr{
				&dst.CallExpr{
					Fun: &dst.Ident{
						Name: middleware,
						Path: "github.com/datadog/orchestrion/instrument",
					},
				},
			},
		},
	}
	markAsInstrumented(stmt)
	return stmt, nil
}
``` 

Tests pass for me locally with this change.

Reply:
Thanks, applied the suggestion!